File: android/sample-ui/src/test/kotlin/com/usesmileid/sampleapps/ui/UseSmileIDSampleSpecTest.kt
package com.usesmileid.sampleapps.ui

import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleProduct
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleScenario
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleThemeScenario
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class UseSmileIDSampleSpecTest {

    private val scenarios: String by lazy { spec("scenarios.json") }

    @Test
    fun flow_scenarios_match_the_spec() {
        val expected = idsOf(scenarios, kind = "flow")
        assertTrue("extracted no flow scenarios", expected.isNotEmpty())
        assertEquals(expected, UseSmileIDSampleScenario.entries.map { it.id })
    }

    @Test
    fun theme_scenarios_match_the_spec() {
        val expected = idsOf(scenarios, kind = "theme")
        assertTrue("extracted no theme scenarios", expected.isNotEmpty())
        assertEquals(expected, UseSmileIDSampleThemeScenario.entries.map { it.id })
    }

    @Test
    fun products_match_the_spec_in_design_order() {
        val expected = PRODUCT_ID.findAll(scenarios).map { it.groupValues[1] }.toList()
        assertTrue("extracted no products", expected.isNotEmpty())
        assertEquals(expected, UseSmileIDSampleProduct.entries.map { it.id })
    }

    /** Four platforms must abbreviate identically: "Enhanced Doc." and "Enh. Doc" are both reasonable. */
    @Test
    fun card_titles_and_families_match_the_spec() {
        val expected = CARD_RUNS.findAll(scenarios).map { it.groupValues[1] to it.groupValues[2] }.toList()
        assertTrue("extracted no card runs", expected.isNotEmpty())
        assertEquals(expected, UseSmileIDSampleProduct.entries.map { it.cardTitle to it.cardFamily })
    }

    /** The mark is one constant, so a card that stops carrying it fails here rather than on a device. */
    @Test
    fun the_smart_selfie_products_carry_the_mark() {
        assertEquals(
            listOf("smartSelfieEnrollment", "smartSelfieAuth"),
            UseSmileIDSampleProduct.entries
                .filter { it.cardFamily == UseSmileIDSampleMarks.SMART_SELFIE }
                .map { it.id },
        )
    }

    @Test
    fun the_captureless_product_is_the_one_the_spec_names() {
        assertEquals(
            listOf("enhancedKyc"),
            UseSmileIDSampleProduct.entries.filterNot { it.capture }.map { it.id },
        )
    }

    /** Scenario entries carry `"id"` then `"kind"`, so the pair is matched together rather than counted. */
    private fun idsOf(json: String, kind: String): List<String> =
        SCENARIO_ID_AND_KIND.findAll(json)
            .filter { it.groupValues[2] == kind }
            .map { it.groupValues[1] }
            .toList()

    private companion object {
        val SCENARIO_ID_AND_KIND =
            Regex("\"id\"\\s*:\\s*\"([A-Za-z]+)\"\\s*,\\s*\"kind\"\\s*:\\s*\"(flow|theme)\"")
        val PRODUCT_ID = Regex("\"id\"\\s*:\\s*\"([A-Za-z]+)\"\\s*,\\s*\"label\"\\s*:")
        val CARD_RUNS =
            Regex("\"cardTitle\"\\s*:\\s*\"([^\"]+)\"\\s*,\\s*\"cardFamily\"\\s*:\\s*\"([^\"]+)\"")
    }
}

File: flutter/sample_ui/test/spec/use_smileid_sample_spec_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:sample_ui/sample_ui.dart';

import 'spec_file.dart';

void main() {
  late Map<String, Object?> scenarios;

  setUpAll(() => scenarios = spec('scenarios.json'));

  List<String> idsOf(String kind) => objects(scenarios['scenarios'])
      .where((Map<String, Object?> it) => it['kind'] == kind)
      .map((Map<String, Object?> it) => it['id']! as String)
      .toList();

  List<Map<String, Object?>> products() {
    final Map<String, Object?> grid =
        scenarios['products']! as Map<String, Object?>;
    return objects(grid['sections'])
        .expand((Map<String, Object?> section) => objects(section['items']))
        .toList();
  }

  test('flow scenarios match the spec', () {
    final List<String> expected = idsOf('flow');
    expect(expected, isNotEmpty, reason: 'extracted no flow scenarios');
    expect(
      UseSmileIDSampleScenario.values.map(
        (UseSmileIDSampleScenario it) => it.id,
      ),
      expected,
    );
  });

  test('theme scenarios match the spec', () {
    final List<String> expected = idsOf('theme');
    expect(expected, isNotEmpty, reason: 'extracted no theme scenarios');
    expect(
      UseSmileIDSampleThemeScenario.values.map(
        (UseSmileIDSampleThemeScenario it) => it.id,
      ),
      expected,
    );
  });

  test('products match the spec in design order', () {
    final List<String> expected = products()
        .map((Map<String, Object?> it) => it['id']! as String)
        .toList();
    expect(expected, isNotEmpty, reason: 'extracted no products');
    expect(
      UseSmileIDSampleProduct.values.map((UseSmileIDSampleProduct it) => it.id),
      expected,
    );
  });

  test('labels, card titles and card families match the spec', () {
    // Four platforms must abbreviate identically: "Enhanced Doc." and "Enh. Doc" are both reasonable.
    expect(
      UseSmileIDSampleProduct.values
          .map(
            (UseSmileIDSampleProduct it) => <String>[
              it.label,
              it.cardTitle,
              it.cardFamily,
            ],
          )
          .toList(),
      products()
          .map(
            (Map<String, Object?> it) => <String>[
              it['label']! as String,
              it['cardTitle']! as String,
              it['cardFamily']! as String,
            ],
          )
          .toList(),
    );
  });

  test('the capture and ID-details flags match the spec', () {
    expect(
      UseSmileIDSampleProduct.values
          .map(
            (UseSmileIDSampleProduct it) => <bool>[
              it.capture,
              it.needsIdDetails,
              it.needsUserDetails,
            ],
          )
          .toList(),
      products()
          .map(
            (Map<String, Object?> it) => <bool>[
              it['capture']! as bool,
              it['needsIdDetails']! as bool,
              it['needsUserDetails']! as bool,
            ],
          )
          .toList(),
    );
  });

  test('the SmartSelfie products carry the mark', () {
    expect(
      UseSmileIDSampleProduct.values
          .where(
            (UseSmileIDSampleProduct it) =>
                it.cardFamily == UseSmileIDSampleMarks.smartSelfie,
          )
          .map((UseSmileIDSampleProduct it) => it.id),
      <String>['smartSelfieEnrollment', 'smartSelfieAuth'],
    );
  });

  test('the captureless product is the one the spec names', () {
    expect(
      UseSmileIDSampleProduct.values
          .where((UseSmileIDSampleProduct it) => !it.capture)
          .map((UseSmileIDSampleProduct it) => it.id),
      <String>['enhancedKyc'],
    );
  });

  test('the two section headings are the ones the ruling settled', () {
    expect(
      UseSmileIDSampleProductSection.values.map(
        (UseSmileIDSampleProductSection it) => it.label,
      ),
      <String>['Authentication', 'Onboarding'],
    );
  });

  test('the four statuses are Title case, not upper-cased', () {
    expect(
      UseSmileIDSampleStatus.values.map(
        (UseSmileIDSampleStatus it) => it.label,
      ),
      <String>['Clear', 'Attention', 'Blocked', 'Processing'],
    );
  });

  test('each status maps onto the feedback role the spec names', () {
    expect(
      UseSmileIDSampleStatus.values.map((UseSmileIDSampleStatus it) => it.role),
      <String>['success', 'warning', 'error', 'info'],
    );
  });
}
